pkg/event/registry.go

```diff
@@ -0,0 +1,142 @@
+package event
+
+import (
+	"github.com/pkg/errors"
+
+	k8sV1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog"
+
+	"github.com/submariner-io/admiral/pkg/log"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+)
+
+type Registry struct {
+	name          string
+	networkPlugin string
+	eventHandlers []Handler
+}
+
+// NewRegistry creates a new registry, the name would normally be the name of the owner, for example
+// 'routeagent', and the networkPlugin must be the network plugin that has been detected (or passed
+// via env-var) this will be used to only keep the relevant handlers in runtime, while discarding
+// the ones that shall not be ran with the system network plugin.
+func NewRegistry(name, networkPlugin string) Registry {
+	return Registry{
+		name:          name,
+		networkPlugin: networkPlugin,
+		eventHandlers: []Handler{},
+	}
+}
+
+func (er *Registry) addHandler(eventHandler Handler) {
+	evNetworkPlugin := eventHandler.GetNetworkPlugin()
+
+	if evNetworkPlugin == AnyNetworkPlugin || evNetworkPlugin == er.networkPlugin {
+		er.eventHandlers = append(er.eventHandlers, eventHandler)
+		klog.Infof("Event handler %q loaded for the %q registry ",
+			eventHandler.GetName(), er.name)
+
+		if err := eventHandler.Init(); err != nil {
+			klog.Fatalf("Event handler %q failed to initialize: %v", eventHandler.GetName(), err)
```

Comment:
Also, it might be better to first initialize and only if it's successful, append it to er.eventHandlers (i.e., move this to a line before 37)

Reply:
Right, we do that everywhere. My thinking was that this kind of failure will be fatal anyway, and we need to stop. So it didn't matter, but for code clarity, when people read it they will see that it can err & exit.

Reply:
It probably is fatal but let the top-level caller decide that. Also it allows the caller(s) to handle fatal appropriately. Remember in the engine we write the GW status on fatal error. Also  propagating the error makes it easier for unit tests.